core/client/fs/src/main/java/alluxio/client/file/InstanceCachingFileSystem.java

```diff
@@ -31,12 +34,21 @@
     super(fs);
     mCache = cache;
     mKey = key;
+    mRefCount = new AtomicInteger(1);
+  }
+
+  /**
+   * Increases the ref count of this instance.
+   */
+  public void incrementCount() {
+    mRefCount.incrementAndGet();
   }
 
   @Override
   public void close() throws IOException {
-    super.close();
-    // TODO(binfan): is this the expected behavior? shouldn't we keep refcount in cache?
-    mCache.remove(mKey);
+    if (mRefCount.decrementAndGet() == 0) {
```

Comment:
Is there a race condition where a client calls close and sees refcount 0, but before calling `mCache.remove` another client gets the filesystem?

Reply:
fixed by serializing all `FileSystemCache.get` and `InstanceCachingFileSystem.close`